[feat] ExportView Audio추출 구현

nakisara01 opened the pull request:
### 설명
ExportView Audio추출 구현


### 관련 이슈

Closes #78 

### 작업 내용

- [x] AudioRenderingService에서 코드&녹음 동시 추출
- [x] intent 내 동작 수정
- [x] NavigationBar Home 이동 동작 추가 

### 스크린샷 (Optional)



### 참고 내용
현재 ExportView의 intent.onAppear에서 sheet, audio 추출을 진행하고 있습니다. 오디오, 시트 추출 각각의 버튼을 눌렀을 때 시트/오디오 추출 함수를 각각 돌리게 하는 방법도 생각해봤지만, 이 경우 버튼을 눌렀을 때 추출 동작이 진행되므로 버튼을 누름과 동시에 ShareLink가 뜨지 않을 가능성이 있어 차라리 ExportView를 로딩할 때 바로 두 추출 함수를 돌리는게 좋을 것 같아 현재와 같이 함수를 구성했습니다. 혹시 이외에 좋은 방법이 생각나시는 분 있다면 코멘트 부탁드리겠습니다. +추가로 테스트해 본 결과 2분정도의 녹음 까지는 0.5-1.0초 이내로 오디오 처리 되는 것으로 확인했습니다

nakisara01 committed:
[#78] feat: add Rendering function at scorePlayer

nakisara01 committed:
[#78] feat: add AudioRenderingService

nakisara01 committed:
[#78] feat: export combined audioFile

nakisara01 committed:
[#78] feat: export name setting

nakisara01 committed:
[#78] fix: refactor AudioRenderingService

nakisara01 committed:
[#78] fix: text Localizable Setting

nakisara01 committed:
Merge branch 'dev' into feat/#78

nakisara01 committed:
[#78] feat: add NavigationBar Action

Segyun commented on GMG/Scenes/Export/ExportView.swift:
```diff
@@ -44,8 +44,17 @@ struct ExportView: View {
+                    Button {
+                        router?.popToRoot()
+                    } label: {
+                        Image("Home")
```
```suggestion
                        Image(.home)
```

Segyun commented on GMG/Scenes/Export/ExportIntent.swift:
onAppear 매개변수로 score를 넘겨주는 방식으로 수정하는 것이 어떨까요?

Segyun commented on GMG/Core/Services/ScoreAudioRenderer.swift:
```diff
@@ -0,0 +1,89 @@
+            throw NSError(domain: "BufferAlloc", code: -1)
+        }
+
+        let totalDuration = score.totalDuration + 0.2
```
0.2는 어떤 값인가요?

nakisara01 replied:
엔진 시작이 딜레이 되는 경우나 offline Rendering은 렌더링 끝 프레임에서 잘림 현상이 발생할 수 있는데, 그래서 0.2초 정도의 여유를 주면 첫 노트 잘림 방지 및 파일이 뚝 끊길 수도 있는 문제를 해결할 수 있어서 0.2를 추가했습니다. 원래는 tailTime으로 정의해서 상수를 만들까 하다가 재사용되는 부분이 없어 그냥 0.2를 더하는 형식으로 바꿨는데 필요하다면 주석 추가하겠습니다

ehdrbdndns replied:
주석 정도만 추가해주면 좋을 것 같스빈다!

Segyun commented on GMG/Core/Models/ScorePlayer/ScorePlayer.swift:
`ScorePlayer`의 프로퍼티를 외부에 노출시키면 캡슐화가 안되어서 외부에서 잘못 변경될 위험이 있습니다.

오디오 렌더링을 `ScorePlayer`의 내부 프로퍼티에 직접 접근하지 않고 구현하는 것이 좋을 것 같습니다.

다음 방식 중 하나로 수정하는 것이 좋을 것 같습니다.

1. 오디오 렌더링 로직을 `ScorePlayer`에 넣기 → ScorePlayer의 역할이 많아짐
2. `ScorePlayer`와 독립적인 `AudioRenderingService`를 구현하기 → ChordCell 관련 Sequencer 로직이 중복됨
3. `ScorePlayer`를 상속받은 `ScoreAudioRenderer` 구현하기 → 불필요한 메서드(play, stop 등)이 남아 있음
4. Score 재생과 렌더링의 공통 로직(준비 로직)만 담은 클래스를 만들어서 `ScorePlayer`와 `ScoreAudioRenderer`가 각각 상속 받아서 구현하기

가장 좋은 것은 두 번째, 네 번째 방법인 것 같긴 한데 작성할 코드가 많아질 것 같아서 세 번째 방법으로 가도 괜찮을 것 같아요!

nakisara01 replied:
코드를 구성하다 보니 궁금한 점이 생겼습니다. Base 클래스를 만들든 상속을 활용하든, private let으로 선언하면 결국 상속받는 쪽에서 사용할 수 없기 때문에 적절한 접근 수준을 다시 고민하게 되더라고요.

그렇다면 internal, fileprivate, 또는 internal private(set) 같은 접근 제어자를 사용해서 상속받은 클래스에서 engine이나 sampler 같은 프로퍼티에 접근해야 할 텐데, 이 중 어떤 방식을 선택하는 것이 적절하다고 생각하시나요?

ehdrbdndns replied:
다른 클래스에서 ScorePlayer에 접근할 때는 프로퍼티를 직접 접근하기보다는, 메소드를 통해 관계를 형성하는 방식이 좋습니다. 결국 중요한 것은 이 관계를 어떻게 정의하느냐입니다. 조이드가 처음에 ScorePlayer를 어떻게 사용하려 했는지 고민하고, 그에 맞는 메소드를 만드는 것이 바람직할 것 같습니다. 또한, internal, fileprivate와 같은 접근 제어자는 현재 상황에서는 필요하지 않을 것 같습니다.

Segyun replied:
Swift에서는 `protected` 상속 받은 클래스 내부에서 접근 가능한 제어자가 없는 것을 몰랐네요!

그렇다면 이렇게 하는 것이 어떨까요? Base 클래스를 만들고 이를 상속 받는 Player와 Renderer를 만들되, Player와 Renderer는 각각 PlayerProtocol과 RendererProtocol을 따르게 만드는 것입니다.

```Swift
class ScoreAudioBase {
    let engine: AVAudioEngine
    ...
}

protocol ScorePlayer {
    func play()
    func stop()
    ...
}

final class DefaultScorePlayer: ScoreAudioBase, ScorePlayer {
    func play() { ... }
    func stop() { ... }
    ...
}

protocol ScoreRenderer {
    func render() -> URL
    ...
}

final class DefaultScoreRenderer: ScoreAudioBase, ScoreRenderer {
    func render() -> URL { ... }
    ...
}

let renderer: ScoreRenderer = DefaultScoreRenderer()
let url = renderer.render() // 🟢: Protocol에 정의된 메서드이므로 가능
renderer.engine.stop() // 🔴: Protocol에 정의되지 않은 프로퍼티이므로 불가능
```

이런 식으로 구현하면 실제로 ScorePlayer 프로토콜로 변수를 만들면 engine과 같은 프로퍼티에 접근하는 것을 막을 수 있습니다.

아니면 프레이가 말씀하신 것처럼 ScorePlayer에 렌더 기능 추가(1번 방법) 하는 것도 크게 나쁘지는 않을 것 같기는 합니다.

Segyun commented on GMG/Core/Services/AudioRenderingService.swift:
```diff
@@ -0,0 +1,89 @@
+                frameCapacity: maxFrames
+            )
+        else {
+            throw NSError(domain: "BufferAlloc", code: -1)
```
`AudioRenderingServiceError`를 선언해서 사용하는 것이 좋을 것 같습니다.

Segyun commented on GMG/Core/Services/AudioRenderingService.swift:
```diff
@@ -0,0 +1,89 @@
+import Accelerate
+import Foundation
+
+final class AudioRenderingService {
```
`Score`만 오디오로 렌더링하는 거라면 `ScoreAudioRenderingService` 혹은 `ScoreAudioRenderer`등의 이름을 사용하는 것이 좋을 것 같아요. 그리고 이니셜라이저가 필요하지 않다면 enum + static 메서드로 사용하는 것이 좋을 것 같습니다.

Segyun commented on GMG/Scenes/Export/ExportIntent.swift:
```diff
@@ -37,7 +37,14 @@ final class ExportIntent: ExportIntentProtocol {
+            let service = AudioRenderingService()
+            do {
+                let renderedURL = try service.renderToAudioFile(
+                    score: score, fileName: "\(score.title).m4a")
```
꼭 수정해야 할 것은 아니긴 한데, 파일 이름이 Score의 타이틀이면 어차피 Score를 전달받으니까 커스텀이 필요할 때만 fileName을 전달 받는 것이 어떨까 싶습니다.

```Swift
func renderToAudioFile(score: Score, fileName: String? = nil) throws -> URL

    ...
    
    let outputURL = FileManager.default.temporaryDirectory.appendingPathComponent("\(fileName ?? score.title).m4a")
```

nakisara01 committed:
[#78] refactor: apply review

- ScoreAudioRender + static func renderToAudioFile로 변경했습니다
- ScoreAudioRenderingServiceError를 선언하여 에러를 호출하도록 변경했습니다.
- onAppear에서 score를 파라미터값으로 사용하도록 변경했습니다.

nakisara01 committed:
[#78] feat: add ScoreAudioEngineBase

ehdrbdndns commented on GMG/Core/Models/ScorePlayer/ScoreAudioEngineBase.swift:
```diff
@@ -0,0 +1,208 @@
+
+        self.audioFile = nil
+
+        self.sampleRate = 48_000
```
48000 수치가 다른 곳에서도 많이 사용되니, constant파일로 빼서 전역 변수처럼 사용하도록 하면 좋을 것 같아요!

ehdrbdndns commented on GMG/Core/Models/ScorePlayer/ScoreAudioEngineBase.swift:
```diff
@@ -0,0 +1,208 @@
+        let chordCells: [ChordCell] = score.retrieveAllChordCells()
+        let audioDuration: TimeInterval = score.totalDuration
+
+        sequencer.rate = 60 / 120
```
60 / 120을 왜 했는지 주석으로 부연 설명해주면 좋을 것 같아요!

Segyun replied:
Sequencer의 기본 속도는 1.0 = 120 bpm이라서 1박에 1초가 되려면 60 / 120으로 해야 됩니다.

ehdrbdndns commented on GMG/Scenes/Export/ExportIntent.swift:
```diff
@@ -15,29 +15,34 @@ final class ExportIntent: ExportIntentProtocol {
+        if let uiImage = renderer.uiImage {
+            model.updateSheetImage(uiImage)
+
+            if let data = uiImage.pngData() {
+                let fileName = "sheet-\(score.id.uuidString).png"
+                let url = FileManager.default
+                    .temporaryDirectory
+                    .appendingPathComponent(fileName)
+
+                do {
+                    try data.write(to: url)
+                    model.updateSheetURL(url)
+                } catch {
+                    print("error")
                 }
```
if let data 보다

guard 구문을 사용하는 것이 현재 context에서는 적합한 것 같습니다.

data가 꼭 필요한 상황인 것 처럼 보여서요!

그리고 에러 구문을 좀 더 구체적으로 적어야할 것 같습니다! 

```suggestion
                guard let data = uiImage.pngData() else { // 에러 시나리오 }

                let fileName = "sheet-\(score.id.uuidString).png"
                let url = FileManager.default
                    .temporaryDirectory
                    .appendingPathComponent(fileName)

                do {
                    try data.write(to: url)
                    model.updateSheetURL(url)
                } catch {
                    // 에러 시나리오
                }
```

nakisara01 replied:
레몬 PR에서 수정한 거 반영하기로 했습니다

ehdrbdndns reviewed (changes requested):
고생하셨습니다 조이드~!

nakisara01 committed:
[#78] fix: seperate ScorePlayer

- ScorePlayer에서 AudioRenderer와 공통으로 사용되는 로직을 ScoreAudioEngineBase로 옮겼습니다
- ScorePlayer, ScoreAudioRenderer에서 ScoreAudioEngineBase에서 상속받아 프로토콜로 사용할 수 있도록 구성했습니다

nakisara01 committed:
Merge branch 'dev' into feat/#78

nakisara01 committed:
[#78] fix: add Constants sampleRate & segmentDur

nakisara01 committed:
[#78] fix: constants err & error message

Segyun commented on GMG/Scenes/Export/ExportIntent.swift:
```diff
@@ -15,29 +15,34 @@ final class ExportIntent: ExportIntentProtocol {
+        }
 
-            model.updateAudioURL(score.audioURL)
+        do {
+            let renderedURL = try DefaultScoreAudioRenderer(score: score)
+                .renderToAudioFile(score: score, fileName: "\(score.title).m4a")
+            model.updateAudioURL(renderedURL)
+        } catch {
+            print("offline render failed: \(error)")
         }
```
Renderer 이니셜라이즈랑 분리하는 것이 어떨까욥

```suggestion
        do {
            let renderer: ScoreAudioRenderer = DefaultScoreAudioRenderer(score: score)
            let renderedURL: URL = try renderer.renderToAudioFile(fileName: "\(score.title).m4a")
            model.updateAudioURL(renderedURL)
        } catch {
            print("offline render failed: \(error)")
}
```

Segyun commented on GMG/Core/Services/ScoreAudioRenderer.swift:
```diff
@@ -0,0 +1,120 @@
+
+    // MARK: - Rendering
+
+    func renderToAudioFile(score: Score, fileName: String? = nil) throws -> URL {
```
```suggestion
    func renderToAudioFile(fileName: String? = nil) throws -> URL {
```

Segyun commented on GMG/Core/Services/ScoreAudioRenderer.swift:
```diff
@@ -0,0 +1,120 @@
+
+        // 출력 파일 URL 설정
+        let outputURL = FileManager.default.temporaryDirectory
+            .appendingPathComponent(fileName ?? score.title)
```
score.title만 넘기면 확장자가 빠져서 `fileName`은 확장자 없이 받는 것이 좋을 것 같아요~

```suggestion
        let outputURL = URL.temporaryDirectory.appending(component: "\(fileName ?? score.title).m4a")
```

nakisara01 committed:
[#78] fix: delete parameter about ScoreAudioRender

nakisara01 committed:
Merge branch 'dev' into feat/#78

nakisara01 committed:
[#78] fix: add file to .m4a

Segyun commented on GMG/Core/Models/ScorePlayer/ScoreAudioEngineBase.swift:
```diff
@@ -0,0 +1,216 @@
+
+extension Tonic.Chord {
+    var midiNoteNumbers: [Int8] {
+        let pitches: [Pitch] = self.pitches(octave: 3)
```
바뀐 옥타브 반영이 안되어 있는데 #115 에서 수정하겠습니다~

Segyun reviewed (approved):
수고하셨습니다~

ehdrbdndns reviewed (approved):
Good 고생하셨습니다 조이드!

ehdrbdndns commented:
Constant 파일이 필요해서 제가 머지 미리 하겠습니다~
